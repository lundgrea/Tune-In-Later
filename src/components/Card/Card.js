import React from 'react'
import './Card.css'
import { connect } from 'react-redux';
// import { toggleFavorite } from '../../actions'


const Card = props => {
  console.log('Card Component props', props.isFavorite)
  const favStar = props.isFavorite ? "https://image.flaticon.com/icons/svg/148/148836.svg" : "https://image.flaticon.com/icons/svg/149/149217.svg"
  return (
    <>
    <h2>{props.artist}</h2>
    <p>{props.albumName}</p>
    <p>{props.genre}</p>
    <img src={props.img} alt=''></img>
    <img onClick={() => props.toggleFavorite(props.id)} className="card__button-fav" src={favStar} alt=''/>
    </>
  )
}

// const mapDispatchToProps = dispatch => ({
//   toggleFavorite: id => dispatch(toggleFavorite(id))
// })

export default Card;