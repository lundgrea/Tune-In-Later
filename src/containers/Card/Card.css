.card__button-fav {
  width: 40px;
}
