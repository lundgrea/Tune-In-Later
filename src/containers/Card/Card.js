import React from 'react'
import { handleErrors, storeFavorites } from '../../actions/index'
import { connect } from 'react-redux'
import { postFavorite, getFavorites, deleteFavorite } from '../../apiCalls/apiCalls'
import './Card.css'



const Card = props => {
  let actionObject
  const favePost = () => {
    postFavorite(props.user.id, albumToPost)
  }


  const albumToPost = {
    album_id: props.album_id,
    artist_name: props.artist_name,
    album_name: props.album_name,
    artwork_url: props.artwork_url,
    release_date: props.release_date,
    content_advisory_rating: props.content_advisory_rating || 'Everybody',
    primary_genre_name: props.primary_genre_name
  }

  const favStar = props.isFavorite ? "https://image.flaticon.com/icons/svg/148/148836.svg" : "https://image.flaticon.com/icons/svg/149/149217.svg"
  if (!props.user){
    actionObject = () => props.handleErrors('Log In To Save Your Albums to Favorites')
  } else {
    actionObject = async () => {
      const userFavorites = await getFavorites(props.user.id)
      const favoritesObject = userFavorites.find(favorite => favorite.album_id === props.album_id)
      if (!favoritesObject) {
        props.toggleFavorite(props.album_id)
        favePost()
      } else{
        deleteFavorite(props.user.id, props.album_id)
    
      }
    }
  }
  
  return (
    <>
    <h2>{props.artist_name}</h2>
    <p>{props.album_name}</p>
    <p>{props.primary_genre_name}</p>
    <img src={props.artwork_url} alt=''></img>
    <img onClick={actionObject} className="card__button-fav" src={favStar} alt=''/>
    </>
  )
}

export const mapStateToProps = store => ({
  favorites: store.favorites
})

export const mapDispatchToProps = dispatch => ({
  handleErrors: (error) => dispatch(handleErrors(error)),
  storeFavorites: (favorites) => dispatch(storeFavorites(favorites))
})

export default connect(null, mapDispatchToProps)(Card);