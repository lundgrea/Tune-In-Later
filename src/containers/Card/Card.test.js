import React from 'react';
import  { shallow } from 'enzyme';
import { Card, mapDispatchToProps, mapStateToProps } from './Card';

describe('Card', () => {
  let wrapper;

  const mockFavorites = [
    {album_name: 'mirrorland', album_id: 12345678, primary_genre_name: 'rap'}
  ];

  const mockUser = {
    id: 1, name: 'aidan', email: 'aidanmckay2000@gmail.com'
  };

  const mockHandleErrors = jest.fn()
  const mockStoreFavorites = jest.fn()
  const mockToggleFavorite = jest.fn()
  beforeEach(() => {
    wrapper = shallow(
      <Card 
        favorites={mockFavorites}
        handleErrors={mockHandleErrors}
        storeFavorites={mockStoreFavorites}
        album_id={1}
        album_name='MirrorLand'
        artist_name='EarthGang'
        artwork_url='asdf'
        content_advisory_rating='explicit'
        isFavorite={true}
        primary_genre_name='Rap'
        release_date='09/05/2019'
        toggleFavorite={mockToggleFavorite}
        user={mockUser}
      />
    )
  })
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot() 
  });

  describe('mapStateToProps', () => {
    it('should return an array of favorite album objects', () => {
      const mockState = {
        favorites: mockFavorites
      }
      const expected = {
        favorites: mockFavorites
      }
      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    })
  });
});

