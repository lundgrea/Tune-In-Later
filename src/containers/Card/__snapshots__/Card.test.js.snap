// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match snapshot 1`] = `
<div
  className="card"
>
  <h2>
    EarthGang
  </h2>
  <p>
    MirrorLand
  </p>
  <p>
    Rap
  </p>
  <img
    alt=""
    src="asdf"
  />
  <img
    alt=""
    className="card__button-fav"
    onClick={[Function]}
    src="https://image.flaticon.com/icons/svg/148/148836.svg"
  />
</div>
`;
